### PR TITLE
Capture logs during testing to log file and retain the log file

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Adds new functionality into the test pipelines to always enable logging. Logs will be generated to `/target/test.log` from the given SDK being tested and will be retained after testing using the pattern `<sdk group>-<sdk-name>-test.log`.

Capturing logging information during testing will help troubleshoot failures during testing, making it easier to fix flaky tests (rather than "just re-run the pipeline").

This change required modifications to `azure-core` testing as it relied on there being no SLF4J logger implementation on the classpath. `ClientLogger`, `DefaultLogger`, and `HttpLoggingPolicy` have updates allowing for a specific logger to be used. This allows for our logging tests to use `DefaultLogger` which now has the ability to control the logging output source (only used in testing at this time).

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: eng/pipelines/templates/steps/build-and-test-native.yml [hunk 1]
```diff
@@ -124,9 +124,11 @@ steps:
     env: ${{ parameters.TestEnvVars }}
     condition: and(succeeded(), eq(variables['TestFromSource'], 'true'))
 
-  - template: ../steps/upload-repository-on-failure.yml
+  - template: /eng/pipelines/templates/steps//upload-repository-on-failure.yml
```
Comment (lmolkova) [inline eng/pipelines/templates/steps/build-and-test-native.yml:127]: nit 
```suggestion
  - template: /eng/pipelines/templates/steps/upload-repository-on-failure.yml
```

**Review by lmolkova — APPROVED** [review]
LGTM!

**Review by srnagar — APPROVED** [review]

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLoggingPolicy.java [hunk 2]
```diff
@@ -560,7 +560,14 @@ private static Integer getRequestRetryCount(Context context) {
     /*
      * Get or create the ClientLogger for the method having its request and response logged.
      */
-    private static ClientLogger getOrCreateMethodLogger(String methodName) {
+    private static ClientLogger getOrCreateMethodLogger(Context context) {
+        ClientLogger logger = (ClientLogger) context.getData("caller-logger").orElse(null);
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLoggingPolicy.java:564]: Probably, but it has been this way for so long, don't know whether anything is using this from outside the SDKs. So, bad name but would keep it for historical purposes (for now)

[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLoggingPolicy.java:564]: After talking offline and viewing this in the larger context.

This is newly added as a way for a full-fledged `ClientLogger` to be passed through the pipeline to be used as the logger here. `client-method` is the old way where a String was passed that gets converted into a `ClientLogger` using that as its name.

For example:

Previously, SwaggerMethodParser would pass the name of the method it represents using `caller-method` and the logic here would capture that and create a `ClientLogger` with that name and cache it.

Now, SwaggerMethodParser can create a `ClientLogger` in its constructor and pass it using `caller-logger`. And this location would no longer need to do caching, which could result in the `ClientLogger` being ejected after a certain amount of time.

Also discussed, this should get renamed to `caller-method-logger` to be consistent with `caller-method`.

**Pull request merged**